File: frontend/src/lib/api/onboarding.ts
import { readPublicErrorMessage } from '$lib/api/errors';
import {
  APPLICATION_UNAVAILABLE_MESSAGE,
  appAvailability,
  fetchWithAvailability,
} from '$lib/stores/app-availability.svelte';

export interface InviteVerifyResponse {
  email: string;
}

export interface OnboardingStateResponse {
  email: string;
}

export interface OnboardingCompleteRequest {
  username: string;
  password: string;
  full_name?: string | null;
  company_name?: string | null;
}

export interface OnboardingCompleteResponse {
  username: string;
  email: string;
}

async function parseJson<T>(res: Response): Promise<T> {
  return res.json() as Promise<T>;
}

async function publicJsonRequest<T>(path: string, init: RequestInit): Promise<T> {
  const response = await fetchWithAvailability(
    path,
    {
      credentials: 'include',
      ...init,
      headers: {
        'Content-Type': 'application/json',
        ...(init.headers || {}),
      },
    },
    true,
  );

  if (!response.ok) {
    if (appAvailability.unavailable) {
      throw new Error(APPLICATION_UNAVAILABLE_MESSAGE);
    }
    throw new Error(await readPublicErrorMessage(response));
  }

  return parseJson<T>(response);
}

export const onboardingApi = {
  verifyInvite: (inviteCode: string) =>
    publicJsonRequest<InviteVerifyResponse>('/api/invites/verify', {
      method: 'POST',
      body: JSON.stringify({ invite_code: inviteCode }),
    }),
  getState: () =>
    publicJsonRequest<OnboardingStateResponse>('/api/onboarding/state', {
      method: 'GET',
      headers: {},
    }),
  complete: (body: OnboardingCompleteRequest) =>
    publicJsonRequest<OnboardingCompleteResponse>('/api/onboarding/complete', {
      method: 'POST',
      body: JSON.stringify(body),
    }),
};

File: backend/app/schemas/onboarding.py
from pydantic import BaseModel, EmailStr, Field, model_validator

from app.services.password_policy import validate_password_strength


class InviteVerifyRequest(BaseModel):
    invite_code: str = Field(min_length=1, max_length=128)


class InviteVerifyResponse(BaseModel):
    email: EmailStr


class OnboardingStateResponse(BaseModel):
    email: EmailStr


class OnboardingCompleteRequest(BaseModel):
    username: str = Field(min_length=3, max_length=100)
    password: str
    full_name: str | None = Field(default=None, max_length=255)
    company_name: str | None = Field(default=None, max_length=255)

    @model_validator(mode="after")
    def _check_password_strength(self):
        validate_password_strength(
            self.password,
            user_inputs=[self.username, self.full_name, self.company_name],
        )
        return self


class OnboardingCompleteResponse(BaseModel):
    username: str
    email: EmailStr
